compiler/rustc_symbol_mangling/src/typeid/typeid_itanium_cxx_abi.rs

```diff
@@ -0,0 +1,418 @@
+            typeid.push_str(&s);
+        }
+
+        ty::Closure(_, substs) | ty::Generator(_, substs, _) => {
+            // u7closureI<return-type><element-type1..element-typeN>E as vendor extended type
+            let mut s = String::from("u7closureI");
+            let sig = substs.as_closure().sig();
+            // Push str without the initial 'F' but with the closing 'E'
+            s.push_str(&(encode_fnsig(tcx, &sig.skip_binder(), dict)?[1..]));
+            compress(dict, (ty, TyQ::None), &mut s);
+            typeid.push_str(&s);
+        }
+
+        // Pointer types
+        ty::Ref(_, ty0, mutbl) => {
+            // [U3mut]u3refI<element-type>E as vendor extended type qualifier and type
+            let mut s = String::new();
+            s.push_str("u3refI");
+            s.push_str(&encode_ty(tcx, *ty0, dict, encode_c_types)?);
+            s.push('E');
+            compress(dict, (ty, TyQ::None), &mut s);
+            if hir::Mutability::Mut == *mutbl {
+                s = format!("{}{}", "U3mut", &s);
+                compress(dict, (ty, TyQ::Mut), &mut s);
+            }
+            typeid.push_str(&s);
+        }
+
+        ty::RawPtr(tm) => {
+            // P[K]<element-type>
+            let mut s = String::new();
+            s.push_str(&encode_ty(tcx, tm.ty, dict, encode_c_types)?);
+            if hir::Mutability::Not == tm.mutbl {
+                s = format!("{}{}", "K", &s);
+                compress(dict, (tm.ty, TyQ::Const), &mut s);
+            };
+            s = format!("{}{}", "P", &s);
+            compress(dict, (ty, TyQ::None), &mut s);
+            typeid.push_str(&s);
+        }
+
+        ty::FnPtr(sig) => {
+            // F<return-type><parameter-type1..parameter-typeN>E
+            let mut s = String::new();
+            s.push_str(&encode_fnsig(tcx, &sig.skip_binder(), dict)?);
+            compress(dict, (ty, TyQ::None), &mut s);
+            typeid.push_str(&s);
+        }
+
+        // Trait types
+        ty::Dynamic(predicates, ..) => {
+            // u3dynI<length><name>E as vendor extended type
+            let mut s = String::from("u3dynI");
+            for predicate in predicates.iter() {
+                match predicate.as_ref().skip_binder() {
+                    ty::ExistentialPredicate::Trait(trait_ref) => {
+                        let name = format!("{}", tcx.opt_item_name(trait_ref.def_id)?);
+                        s.push_str(&format!("{}", name.len()));
+                        s.push_str(&name);
+                    }
+                    ty::ExistentialPredicate::Projection(_) => {}
+                    ty::ExistentialPredicate::AutoTrait(_) => {}
+                };
+            }
+            s.push('E');
+            compress(dict, (ty, TyQ::None), &mut s);
+            typeid.push_str(&s);
+        }
+
+        // Placeholder types
+        ty::Bound(..) | ty::Error(_) | ty::Infer(..) | ty::Param(_) | ty::Placeholder(_) => {
```

Comment:
I believe these should be a `bug!()` too.

Reply:
Done. Do you know if we can get the variant name dynamically?

Reply:
What do you mean?

Reply:
The name of the variant that matched to display in the error message (e.g., `ty::Bound`, `ty::Error`, `ty::Infer`, etc.).

Reply:
You could use `bug!("encode_ty: unexpected `ty::Ty` variant {:?}", ty.kind()` I think.

Reply:
Done.